lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
+								node = childNode,
+								toElement = newChildElement,
+							}))
+						end
+					end
 
-		local oldRef = oldElement.props[Core.Ref]
-		local newRef = newElement.props[Core.Ref]
+					for key in pairs(oldValue) do
+						local newChildElement = newValue[key]
+						local childNode = node.children[key]
 
-		-- Change the ref in one pass before applying any changes.
-		-- Roact doesn't provide any guarantees with regards to the sequencing
-		-- between refs and other changes in the commit phase.
-		if newRef ~= oldRef then
-			applyRef(oldRef, nil)
-			applyRef(newRef, instanceHandle._rbx)
+						if newChildElement == nil then
+							DEBUG_print("\t\tScheduling unmount of child", key)
+							scheduleTask(tree, taskUnmountNode({
+								node = childNode,
+							}))
+						end
+					end
+				else
+					DEBUG_print("\tSetting", prop, newValue)
+					node.rbx[prop] = newValue
+				end
+			end
 		end
 
-		-- Update properties and children of the Roblox object.
-		Reconciler._reconcilePrimitiveProps(oldElement, newElement, instanceHandle._rbx)
-		Reconciler._reconcilePrimitiveChildren(instanceHandle, newElement)
-
-		instanceHandle._element = newElement
-
-		return instanceHandle
-	elseif isFunctionalElement(newElement) then
-		instanceHandle._element = newElement
-
-		local rendered = newElement.component(newElement.props)
-		local newChild
-
-		if instanceHandle._child then
-			-- Transition from tree to tree, even if 'rendered' is nil
-			newChild = Reconciler._reconcileInternal(instanceHandle._child, rendered)
-		elseif rendered then
-			-- Transition from nil to new tree
-			newChild = Reconciler._mountInternal(
-				rendered,
-				instanceHandle._parent,
-				instanceHandle._key,
-				instanceHandle._context
-			)
+		for prop in pairs(fromElement.props) do
+			if not visitedProps[prop] then
+				-- TODO: Use real mechanism for setting default values
+				node.rbx[prop] = nil
+			end
 		end
+	end
+end
 
-		instanceHandle._child = newChild
+local function processTreeTasksAsync(tree, timeBudget)
+	if #tree.tasks == 0 then
+		return
+	end
 
-		return instanceHandle
-	elseif isStatefulElement(newElement) then
-		instanceHandle._element = newElement
+	DEBUG_warn(("-"):rep(80))
+	DEBUG_print("Async frame:", #tree.tasks, "tasks in queue; marker at", tree.taskIndex)
 
-		-- Stateful elements can take care of themselves.
-		instanceHandle._instance:_update(newElement.props)
+	tree.tasksRunning = true
 
-		return instanceHandle
-	elseif isPortal(newElement) then
-		if instanceHandle._rbx ~= newElement.props.target then
-			local parent = instanceHandle._parent
-			local key = instanceHandle._key
-			local context = instanceHandle._context
+	local startTime = tick()
+	local i = tree.taskIndex
+	while true do
+		local task = tree.tasks[i]
 
-			Reconciler.unmount(instanceHandle)
+		local totalTimeElapsed = tick() - startTime
 
-			local newInstance = Reconciler._mountInternal(newElement, parent, key, context)
+		if task == nil then
```

Comment:
The goal here was to minimize the amount of time spent managing tasks when running them to try to keep as much time for doing actual work as possible.

The issue is that moving elements from the front of a list is super expensive (you have to shift everything down); we could probably use a deque here instead, which would abstract away the `taskIndex` stuff and probably lead to cleaner code.

Reply:
I don't think I explained my idea clear enough. It wont increase the overhead of the task management system by much at all, but will prevent lua's heap from growing out of control. In processTreeTasks we can insert this code snippet near the top:
```lua
tree.tasksRunning = true
tree.runningTasks = tasks
tree.tasks = {}
```
and change references to tree.tasks in the function to tree.runningTasks. In the check to see if we have run out of tasks:
```lua
if task == nil then
    if tree.tasks[1] == nil then
        break
    else
        tree.runningTasks = tree.tasks
        tree.tasks = {}
    end
end
```

That way we can let the garbage collector do its job while we are running and don't gobble up memory. The added overhead is a reference reassignment, and perhaps a couple table creations. The advantage is that the garbage collector can do its job and keep us performing optimally, in addition to avoiding the scenario where a task always creates another task and eventually crashing the system because no memory can be freed.